Make location of TrackInfoObject immutable

uklotzde opened the pull request:
The location of a TrackInfoObject is immutable, even if the existing code suggested something else. I also started to align the constructors of TrackInfoObject and removed redundant code.

A small cleanup PR for TrackInfoObject, more will follow.

uklotzde committed:
Delete redundant constructor

uklotzde committed:
Delete redundant QFileInfo instance

The location of a track is always calculated by obtaining the absolute
file path from QFileInfo.

uklotzde committed:
Make the track location immutable

uklotzde committed:
Delete unnecessary destructor

uklotzde committed:
Initialize members directly and reuse initialize()

uklotzde committed:
Fix signature of operator<< for QDebug

uklotzde committed:
Log full track location instead of only the file name

uklotzde committed:
Rename file name/size functions in TrackInfoObject

daschuer commented:
The failed Travis build seams to be a false positive due to a timeout.

daschuer commented on src/library/basesqltablemodel.cpp:
```diff
@@ -842,7 +842,12 @@ void BaseSqlTableModel::setTrackValueForColumn(TrackPointer pTrack, int column,
+                    << pTrack->getLocation()
+                    << "to"
+                    << value.toString();
+        }
```
IMHO we can be more strict here and assert if one tries to modify the location.
DEBUG_ASSERT_AND_HANDLE(false) 
This would be called if a user edits the location column. 
Why would a user do that?

uklotzde replied:
I will add an else block with a DEBUG_ASSERT_AND_HANDLE macro that prints a 
warning when trying to edit an unsupported column:
- COLUMN_LIBRARYTABLE_LOCATION
- COLUMN_LIBRARYTABLE_FILETYPE
- COLUMN_LIBRARYTABLE_DURATION
- COLUMN_LIBRARYTABLE_BITRATE
- ...and all other columns that are not handled in the if/else before

On 09/20/2015 06:06 PM, Daniel Schürmann wrote:
>
> In src/library/basesqltablemodel.cpp 
> <https://github.com/mixxxdj/mixxx/pull/718#discussion_r39931353>:
>
> > @@ -842,7 +842,12 @@ void BaseSqlTableModel::setTrackValueForColumn(TrackPointer pTrack, int column,
> >      } else if (fieldIndex(ColumnCache::COLUMN_LIBRARYTABLE_TRACKNUMBER) == column) {
> >          pTrack->setTrackNumber(value.toString());
> >      } else if (fieldIndex(ColumnCache::COLUMN_LIBRARYTABLE_LOCATION) == column) {
> > -        pTrack->setLocation(value.toString());
> > +        DEBUG_ASSERT_AND_HANDLE(pTrack->getLocation() == value.toString()) {
> > +            qWarning() << "The location of track cannot be modified from"
> > +                    << pTrack->getLocation()
> > +                    << "to"
> > +                    << value.toString();
> > +        }
>
> IMHO we can be more strict here and assert if one tries to modify the 
> location.
> DEBUG_ASSERT_AND_HANDLE(false)
> This would be called if a user edits the location column.
> Why would a user do that?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/718/files#r39931353>.
>

daschuer commented on src/trackinfoobject.cpp:
```diff
@@ -43,40 +43,37 @@
+        QString filename = XmlParse::selectNodeQString(nodeHeader, "Filename");
+        QString location = QDir(XmlParse::selectNodeQString(nodeHeader, "Filepath")).filePath(filename);
+        return QFileInfo(location);
+    }
```
We are probably not able to test a 1.7 upgrade. 
IMHO we should issue a warning, that a 1.7 library is too old and that the user should install Mixxx 1.12 fist, which still includes the upgrade code. 

Any thought?

uklotzde replied:
Then let's rid of this legacy code entirely!

No one is able to test it and you could use one of the versions up to 1.12 
if you need to migrate a library from 1.7.

On 09/20/2015 06:11 PM, Daniel Schürmann wrote:
>
> In src/trackinfoobject.cpp 
> <https://github.com/mixxxdj/mixxx/pull/718#discussion_r39931406>:
>
> >      initialize(parseHeader, parseCoverArt);
> >  }
> >
> > +namespace {
> > +    QFileInfo parseFileInfo(const QDomNode& nodeHeader) {
> > +        QString filename = XmlParse::selectNodeQString(nodeHeader, "Filename");
> > +        QString location = QDir(XmlParse::selectNodeQString(nodeHeader, "Filepath")).filePath(filename);
> > +        return QFileInfo(location);
> > +    }
>
> We are probably not able to test a 1.7 upgrade.
> IMHO we should issue a warning, that a 1.7 library is too old and that the 
> user should install Mixxx 1.12 fist, which still includes the upgrade code.
>
> Any thought?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/718/files#r39931406>.
>

daschuer replied:
@rryan: Do you agree, that it is time to drop Mixxx 1.7 upgrade feature?

daschuer commented on src/trackinfoobject.h:
```diff
@@ -98,18 +92,24 @@ class TrackInfoObject : public QObject {
 
+    QFileInfo getFileInfo() const {
+        // Copying a QFileInfo is thread-safe (implicit sharing), no locking needed.
+        return m_fileInfo;
```
Great!

daschuer commented on src/trackinfoobject.h:
```diff
@@ -98,18 +92,24 @@ class TrackInfoObject : public QObject {
+    // Returns the size of the file in bytes
+    int getFileSize() const;
     // Returns whether the file exists on disk or not. Updated as of the time
     // the TrackInfoObject is created, or when setLocation() is called.
```
Outdated comment

uklotzde committed:
Exclude certain columns from editing in table model

uklotzde committed:
Delete outdated comment

uklotzde committed:
Get rid of LegacyLibraryImporter

A Mixxx library from a version prior to 1.8 must be migrated by using
one of the versions 1.8/1.9/1.10/1.11/1.12

uklotzde committed:
Move code from initialize() into constructor

uklotzde committed:
Add more comments about thread-safety of QFileInfo

daschuer commented:
LGTM. Though I would like to have a third opinion about 1.7 drop. 

All: Any thoughts?

uklotzde committed:
Delete obsolete file headers and redundant #include directives

uklotzde committed:
Reorder member functions in class definition

uklotzde committed:
Remove TrackPointer from cover art widgets

...no need to keep a reference of TrackInfoObject.

uklotzde committed:
Merge branch 'master' of https://github.com/mixxxdj/mixxx.git into TrackLocationImmutable

uklotzde committed:
Move code for obtaining the global key as text to KeyUtils

uklotzde committed:
Merge branch 'master' into TrackLocationImmutable

daschuer commented:
LGTM. Thank you. 
If there are no objections against dropping 1.7 library support, I will merge this soon.
